### PR TITLE
Draft EIP: BEGINDATA opcode

### PR DESCRIPTION
**Pull request by MrChico**
Waiting for EIP number to be assigned. Will post link to discussions after having opened this PR.

### CODE REVIEW

**Comment by MrChico** [comment]
Discussions-to: https://ethereum-magicians.org/t/new-opcode-begindata/3727

File: EIPS/eip-2327.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+---
+eip: 2327
+title: BEGINDATA opcode
+author: Martin Lundfall martin.lundfall@protonmail.com
+discussions-to: https://ethereum-magicians.org/t/new-opcode-begindata/3727
+status: Draft
+type: Standards Track
+category Core
```
Comment (shemnon) [inline EIPS/eip-2327.md]:
```suggestion
category: Core
```

File: EIPS/eip-2327.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+---
+eip: 2327
+title: BEGINDATA opcode
+author: Martin Lundfall martin.lundfall@protonmail.com
```
Comment (axic) [inline EIPS/eip-2327.md]: Can you fix the email to be encapsulated in `<>`? You can also use github username instead as `(@MrChico)`.

File: EIPS/eip-2327.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+The `BEGINDATA` opcode has been suggested before as part of the EIP `Subroutines and Static Jumps for the EVM` [EIP-615](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-615.md) as a way to determine the position of jumptables in contract bytecode. It is here introduced in its own right in order to exclude data from the JUMPDEST analysis of contracts, making it impossible to jump to data. This makes it easier for static analysis tools to analyse contracts, allows disassemblers, chain explorers and debuggers to not display data as a mess of INVALID opcodes and may even provide a marginal improvement in performance. Additionally, it paves the way for suggestions such as [EIP 1712](https://github.com/ethereum/EIPs/pull/1712) to disallow unused opcodes, jumptables [EIP-615](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-615.md) and speculative proposals to disallow for deployment of contracts with stack usage violations.
```
Comment (axic) [inline EIPS/eip-2327.md]: Please use the canonical URL (https://eips.ethereum.org/EIPS/eip-XX) whenever possible.

File: EIPS/eip-2327.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+While computing the valid JUMPDESTs of a contract, halt analysis if `BEGINDATA` is encountered. A `JUMP` to a value higher than the `pc` value of `BEGINDATA` should throw with a `BAD_JUMP_DESTINATION` error.
+If `BEGINDATA` is encountered during contract execution, it has the same semantics as `STOP`. It uses 0 gas.
+
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
+The byte `0xb6` was chosen to align with [EIP-615](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-615.md). 
+The choice to STOP if `BEGINDATA` is encountered is somewhat arbitrary. An alternative would be to `THROW`.
```
Comment (axic) [inline EIPS/eip-2327.md]: There is no such thing as "throw" defined anywhere.

Comment (axic) [inline EIPS/eip-2327.md]: It may make it nicer to read if all JUMPDEST and STOP opcodes are escaped.

File: EIPS/eip-2327.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+## Backwards Compatibility
+<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->
+The proposal will not change any existing contracts unless their current behaviour relies upon the usage of unused opcodes.
+
+## Test Cases
+<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->
+Test cases should include:
+1) A contract which jumps to a destination `X`, where `X` has a pc value higher than the `BEGINDATA` opcode, and the byte at `X` is `0x5b`. This should fail with a `BAD_JUMPDEST`.
```
Comment (axic) [inline EIPS/eip-2327.md]: Please use BAD_JUMPDEST or BAD_JUMP_DESTINATION consistently.

Comment (axic) [reply EIPS/eip-2327.md]: Albeit there is no such error clearly defined in EVM semantics, I'd still opt to keep that in the EIP (for now).

Comment (MrChico) [reply EIPS/eip-2327.md]: I am referring to the status codes as defined by EVMC https://evmc.ethereum.org/group__EVMC.html#ga4c0be97f333c050ff45321fcaa34d920. Even if they haven't been universally accepted, they do provide the most systematic treatment of exceptions I'm aware of.

Comment (axic) [reply EIPS/eip-2327.md]: I'm aware of them and glad to see them utilised, they haven't been referenced yet (apart from #2003).

**Comment by MrChico** [comment]
Thanks @axic for the review

**Pull request merged**